### PR TITLE
Non-record: PR #1855 + MP3 marker-pair fusion + alias smear boundary — val_bpb 1.05858 (3-seed mean, DGX canonical)

### PR DESCRIPTION
**Pull request by izlley**
# PR #1855 stack + MP3 marker-pair fusion + alias smear boundary

**val_bpb: 1.05858203** (3-seed mean: 42, 0, 314; phased TTT, canonical 50 K-val-doc on DGX) | size_max 15,906,335 bytes | 8×H100 80 GB SXM, 600 s wallclock

Beats PR #1855's 3-seed mean **1.06108** by **−0.00250 BPB** (3-seed mean).

> **Update (post-deadline correction)**: numbers below replace the originally submitted runpod runs (val_tokens 36,562,944) with a strict canonical reproduction on DGX (val_tokens 43,767,808). Same `train_gpt.py` (md5 verified), same seeds, same hparams. See the comment thread for the val-partition root-cause + correction story.

## Background

This submission grew out of a week of solo experiments on a smaller
in-house stack that plateaued around val_bpb 1.065. To break past that,
I rebased onto PR #1855 (currently the strongest published candidate)
and added two orthogonal contributions: **MP3 marker-pair fusion**
(data-side) and **alias smear boundary** (mechanism-side).

## What this submission adds

Two additions on top of [PR #1855](https://github.com/openai/parameter-golf/pull/1855)
(11L XSA + LQER asym int4 + SparseAttnGate + BOS-fixed SmearGate + Polar-Express
MuonEqR + fused softcapped CE Triton kernel + GPTQ int6 + int7 embed + per-row
int8 attn-gate + per-group lrzip+brotli + phased TTT + 9-hparam stack;
3-seed mean **1.06108**):

1. **MP3 marker-pair fusion** (vocab surgery) — the three 2-grams
   `[▁, TITLE]` / `[▁, ALLCAPS]` / `[▁, CAPNEXT]` are fused into single alias
   donor tokens (donors 8/9/10 from byte-fallback IDs that occur 0× in the
   CaseOps corpus). Word X preserved → no d=1 prediction collapse.
   **Token saving 8.47 %** on train + val streams.

2. **Alias smear boundary** — at positions immediately following an alias
   token, SmearGate's previous-position contribution is fully disabled
   (`ALIAS_PREV_SMEAR_SCALE=0.0`). Regular non-alias positions are unchanged.

## The single conceptual rule

```
[regular → regular]: SmearGate normal (PR #1855 default)
[regular → alias  ]: SmearGate normal at the alias position itself
[alias   → next   ]: SmearGate fully OFF
```

Alias donors already encode dense `(▁ + marker)` signal in a single position;
disabling SmearGate at the next position prevents over-injecting the same
information twice. No arbitrary dampening constant.

## Code-level patch

**+62 / −2 lines** against PR #1855's `train_gpt.py`, organised in 5 hunks:
- `__init__` flag + `alias_is_alias` buffer
- per-position multiplicative mask in `_forward_hidden`
- same mask in TTT path `forward_ttt`
- `alias_map.json` load + buffer re-registration in `deserialize`
- alias rows warm-init + buffer activation in `train_model`

`ALIAS_PREV_SMEAR_SCALE=0.0` is set in `run_3seed.sh` (code default 0.25 overridden so the boundary rule is unambiguous).

## Per-seed results

| Seed | val_bpb (phased TTT) | size (bytes) | steps | step_avg |
|---|---|---|---|---|
| 42 | **1.05840959** | 15,898,576 | 5041 | 118.96 ms |
| 0 | **1.05841119** | 15,906,335 | 5034 | 119.10 ms |
| 314 | **1.05892531** | 15,904,260 | 5040 | 118.97 ms |
| **3-seed mean (42, 0, 314)** | **1.05858203** | 15,906,335 (max) | 5038 | 119.01 ms |

For reference, the originally submitted runpod 3-seed mean was 1.05917838 (same method, smaller non-canonical val partition).

## Why this works

- **Token saving (training-side)**: ~14.97 B → ~13.71 B train tokens (8.47% reduction); same 600 s budget sees more *unique* docs per step. Bytes lossless: val sidecar byte sum unchanged.
- **NLL distance breakdown (eval-side)**: d=1 is preserved (word X unchanged) and d=4+ shows compounding wins (-1.2 ~ -1.7 mbpb per bucket); alias positions themselves cost +1.4 mbpb but the non-alias gains dominate. Net **-1.46 mbpb** at int6 level.

## Pipeline

See [`README.md`](./README.md) for full reproduction steps. Brief:

```bash
apt-get install -y lrzip
pip install -r requirements.txt
pip install --no-deps flash_attn_3 --find-links https://windreamer.github.io/flash-attention3-wheels/cu128_torch291/
python3 download_docs.py                         # raw FineWeb-10B docs (~45 GiB)
python3 prepare_caseops_data.py --docs ./data/datasets/docs_selected.jsonl --out ./data --sp tokenizers/<sp>.model
python3 prepare_marker_pair_v3.py                # MP3 vocab surgery
bash run_3seed.sh                                # 3-seed training
```

## Files

- `train_gpt.py` — PR #1855 + 5-hunk MP3 patch
- `prepare_caseops_data.py` — multiprocess CaseOps tokenizer
- `prepare_marker_pair_v3.py` — MP3 vocab surgery (per-shard parallel)
- `download_docs.py` — minimal HF Hub wrapper
- `lossless_caps.py`, `tokenizers/...model`, `alias_map.json`, `requirements.txt`, `run_3seed.sh`
- `train_seed{42,0,314}.log` — per-seed run logs (DGX corrected canonical reproduction)

## Hardware

8 × H100 80 GB SXM (DGX), 600 s wallclock per seed. Toolchain: driver 535.247.01, Python 3.12.3, PyTorch 2.9.1+cu128, CUDA 12.8, cuDNN 91002, flash_attn_3 3.0.0+20260303, flash_attn 2.8.3, lrzip 0.660.

### CODE REVIEW

**Comment by izlley** [comment]
## Post-deadline correction: canonical val partition + DGX 3-seed reproduction

**TL;DR**: After the deadline I discovered my submitted runs used a non-canonical val partition. I re-ran the **same code with the same seeds and same hparams** on a strictly canonical 50K-val-doc dataset (rebuilt single-pass on my local DGX) and updated the README, submission.json, and the three reported `train_seed*.log` artifacts to reflect the corrected reproduction.

**Corrected 3-seed mean: `val_bpb = 1.05858203`** (Δ vs PR #1855's 1.06108: **−2.50 mbpb**).

---

### What was wrong

My originally submitted logs report `val_tokens = 36,562,944` (≈41,597 val docs after MP3). The canonical setup `prepare_caseops_data.py --val-docs 50000` (single-process) produces `val_tokens = 43,767,808` (≈50,000 val docs after MP3).

**Root cause** — I was under data-build deadline pressure on runpod, so I sliced the source jsonl into four parallel `prepare_caseops_data.py` jobs (`--skip-docs/--max-docs`) to fit the time budget. Each parallel chunk independently emitted *its own* first-N as val. The union was therefore not the strict global first-50K val partition the challenge specifies. I did not catch the mismatch until after the deadline because the chunked val set still produced a sensible val_bpb signal during training.

I deeply apologize for the post-hoc correction. It was unintended noise from my deadline-pressure prep pipeline, not a methodology choice.

### How I recomputed

- Rebuilt the dataset on my DGX with **single-process** `prepare_caseops_data.py --val-docs 50000` from the canonical `willdepueoai/parameter-golf` source jsonl.
- Re-ran the **identical** `train_gpt.py` already in this PR (`md5 f4a81ac75c987126f330ee30f888ee57`, byte-verified).
- **Identical** seeds: 42, 0, 314.
- **Identical** hparams: full PR #1855 reproduce env + `MARKER_PAIR_MODE=1` + `MARKER_PAIR_W_SPACE=0.4` + `MARKER_PAIR_W_TITLE=0.6` + `ALIAS_PREV_SMEAR_SCALE=0.0`. Only deviation: did **not** set `PYTORCH_CUDA_ALLOC_CONF=expandable_segments:True` (a memory-allocator hint that does not affect numerics; it had been added on runpod during a phased-TTT OOM retry on seed 42 alone).
- 8×H100 80 GB SXM (DGX), torch 2.9.1+cu128, FA3, lrzip 0.660 — same toolchain as the runpod submission.

### Results

| | runpod (originally submitted) | DGX (corrected canonical) |
|---|---|---|
| `val_tokens` | 36,562,944 | **43,767,808** (canonical 50 K val docs) |
| seed 42 `val_bpb` | 1.05889359 | **1.05840959** |
| seed 0 `val_bpb` | 1.05961546 | **1.05841119** |
| seed 314 `val_bpb` | 1.05902610 | **1.05892531** |
| **3-seed mean** | 1.05917838 | **1.05858203** |
| std (sample) | 0.00031 | **0.00030** |
| size_max (bytes) | 15,907,150 | **15,906,335** |

The PR ranks the same; only the val partition is now strict canonical. The methodology, architecture, and reported method (PR #1855 stack + MP3 marker-pair fusion + alias smear boundary) are unchanged from the original submission.

### Files updated in this commit

- `train_seed42.log`, `train_seed0.log`, `train_seed314.log` — replaced with the DGX canonical-val runs.
- `submission.json` — DGX numbers + grouped `env` field (toolchain versions).
- `README.md` — headline, per-seed table, MP3 contribution measurement, and surrounding prose updated to the DGX numbers.

PR title also updated to reflect the corrected mean. Thank you for your patience with this corrigendum — please consider the DGX 3-seed mean as the canonical, methodology-faithful reproduction of this submission.

**Comment by cocohearts** [comment]
Leaderboard audit note (pre-cutoff state): I don't think this is valid as a record row before the cutoff. The latest pre-cutoff state used a non-canonical validation partition, and the canonical correction/evidence was pushed after the 2026-05-02T00:00:00Z cutoff. I would not count the post-cutoff correction for the cutoff-limited leaderboard update.

**Comment by izlley** [comment]
> Leaderboard audit note (pre-cutoff state): I don't think this is valid as a record row before the cutoff. The latest pre-cutoff state used a non-canonical validation partition, and the canonical correction/evidence was pushed after the 2026-05-02T00:00:00Z cutoff. I would not count the post-cutoff correction for the cutoff-limited leaderboard update.

Thanks for the audit. You're right — the pre-cutoff state used a non-canonical val partition, and the canonical numbers landed post-cutoff. Happy to drop the record claim — would love to get this merged as a non-record so the methodology (PR#1855 + MP3 marker-pair fusion + alias smear boundary) is documented for others.
